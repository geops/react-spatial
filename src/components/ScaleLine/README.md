#

This demonstrates the use of ScaleLine.

```js
import React, { Component } from  'react';
import BasicMap from 'react-spatial/components/BasicMap';
import ConfigReader from 'react-spatial/ConfigReader';
import OLMap from 'ol/Map';
import ScaleLine from 'react-spatial/components/ScaleLine';

class ScaleLineExample extends Component {
  constructor(props) {
    const layerConf = [{
      name: 'OSM Baselayer',
      visible: true,
      data: {
        type: 'xyz',
        url: 'https://c.tile.openstreetmap.org/{z}/{x}/{y}.png',
      },
    }];

    this.map = new OLMap();
    this.layers = ConfigReader.readConfig(layerConf);
  }

  render() {
    return (
      <div style={{position:'relative'}}>
        <BasicMap
          map={this.map}
          zoom={3}
          layers={this.layers}
          tabIndex={0}
        />
        <ScaleLine map={this.map} />
      </div>
    );
  }
}

<ScaleLineExample />
```
