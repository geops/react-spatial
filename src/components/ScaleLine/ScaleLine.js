import React, { Component } from 'react';
import PropTypes from 'prop-types';
import OLScaleLine from 'ol/control/ScaleLine';
import OLMap from 'ol/Map';

const propTypes = {
  /**
   * Openlayers map.
   */
  map: PropTypes.instanceOf(OLMap).isRequired,

  /**
   * CSS class for the container.
   */
  className: PropTypes.string,

  /**
   * Options for ol/control/ScaleLine.
   * See https://openlayers.org/en/latest/apidoc/module-ol_control_ScaleLine-ScaleLine.html
   */
  options: PropTypes.shape(),
};

const defaultProps = {
  className: 'tm-scale-line',
  options: {},
};

class ScaleLine extends Component {
  constructor(props) {
    super(props);
    this.ref = React.createRef();
  }

  componentDidMount() {
    const { map, options } = this.props;

    this.control = new OLScaleLine({
      ...options,
      ...{ target: this.ref.current },
    });

    map.addControl(this.control);
  }

  componentWillUnmount() {
    const { map } = this.props;
    map.removeControl(this.control);
  }

  render() {
    const { className } = this.props;
    return <div className={className} ref={this.ref} />;
  }
}

ScaleLine.propTypes = propTypes;
ScaleLine.defaultProps = defaultProps;

export default ScaleLine;
