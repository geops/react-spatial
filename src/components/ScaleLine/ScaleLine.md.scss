.tm-scale-line-example {
  position: relative;
}
