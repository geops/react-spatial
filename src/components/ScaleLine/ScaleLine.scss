.tm-scale-line .ol-scale-line-inner {
  border-width: 0 1px 1px 1px;
  border-color: 1px solid $gray;
  border-style: solid;
  padding: $padding-base;
  font-size: $font-size-small;
}
