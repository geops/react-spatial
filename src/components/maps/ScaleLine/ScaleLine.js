import React, { useRef, useEffect } from 'react';
import PropTypes from 'prop-types';
import OLScaleLine from 'ol/control/ScaleLine';
import OLMap from 'ol/Map';

const propTypes = {
  /**
   * Openlayers map.
   */
  map: PropTypes.instanceOf(OLMap).isRequired,

  /**
   * Options for ol/control/ScaleLine.
   * See https://openlayers.org/en/latest/apidoc/module-ol_control_ScaleLine-ScaleLine.html
   */
  options: PropTypes.shape(),
};

const defaultProps = {
  options: {},
};

function ScaleLine({ map, options, ...other }) {
  const ref = useRef();

  useEffect(() => {
    const control = new OLScaleLine({
      ...options,
      ...{ target: ref.current },
    });

    map.addControl(control);
    return () => {
      map.removeControl(control);
    };
  }, [map]);

  // eslint-disable-next-line react/jsx-props-no-spreading
  return <div className="rs-scale-line" ref={ref} {...other} />;
}

ScaleLine.propTypes = propTypes;
ScaleLine.defaultProps = defaultProps;

export default React.memo(ScaleLine);
