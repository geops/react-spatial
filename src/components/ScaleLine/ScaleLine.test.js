import 'jest-canvas-mock';
import React from 'react';
import { configure } from 'enzyme';
import OLMap from 'ol/Map';
import OLView from 'ol/View';
import Adapter from 'enzyme-adapter-react-16';
import renderer from 'react-test-renderer';
import ScaleLine from './ScaleLine';

configure({ adapter: new Adapter() });

describe('ScaleLine', () => {
  test('matches snapshot', () => {
    const map = new OLMap({ view: new OLView({ zoom: 7, center: [0, 0] }) });
    const component = renderer.create(<ScaleLine map={map} />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
