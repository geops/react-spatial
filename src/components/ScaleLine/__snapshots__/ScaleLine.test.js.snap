// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ScaleLine matches snapshot 1`] = `
<div
  className="tm-scale-line"
/>
`;
